include/circt/Dialect/HW/HWStructure.td

```diff
@@ -203,6 +203,12 @@ def HWModuleOp : HWModuleOpBase<"module",
     LogicalResult verifyBody();
   }];
 
+  let extraClassDefinition = [{
```

Comment:
Kind of... The main problem with this is (as far as I can tell), a `let` is an override which replaces the old value. The trick of doing `let extraClassDefinition = extraClassDefinition # [{ ... }];` does not seem to work.

What I can do here is put the default implementation into a `code` block inside `HWModuleOp` and then re-use that in multiple places. Something like:

```
class Foo {
  code getNumPorts = [{ ... }];
  code otherMethod = [{ ... }];
}

def Bar : Foo {
  let extraClassDefinition = getNumPorts # otherMethod # [{
    ... 
  }];
}
```

I've used this trick before with Grand Central-related attributes where each subtype has some mixture of methods. See: https://github.com/llvm/circt/blob/main/include/circt/Dialect/FIRRTL/FIRRTLAttributes.td#L61

WDYT?

Reply:
> Kind of... The main problem with this is (as far as I can tell), a `let` is an override which replaces the old value. The trick of doing `let extraClassDefinition = extraClassDefinition # [{ ... }];` does not seem to work.

I think this is why the existing `HWModuleBaseOp` does `let extraClassDeclaration = extraModuleClassDeclaration # [{ ... }];`  (invents new variable for users of the class to define to append to shared class declarations), unless that doesn't work either.

But the main point was only to try to avoid repetition as more gets moved over, however it's done.

Reply:
I wasn't grokking that this was creating the new `extraModeulClassDeclaration` and was reading that as `let foo = foo # [{ ... }];` not `let foo = bar # [{ ... }];`. I think that makes sense and it is cleaner. I can switch this (and the latter commit that needs the same thing to use this approach).

Reply:
Ah, didn't see that later commit was building on this already, okay then yeah I guess getting something we like early/here is good!

Thanks!

Reply:
This looks like it works great. The one difference seems to be that this has to provide a default (empty) table gen string in order for it to work. As in:

```
code foo = [{}];

let bar = foo # [{
  stuff
}];
```

The formulation of `?` will cause an error if there is no let to define it. This doesn't come up for the `code extraModuleClassDeclaration` as this is defined in every sub-class.

I'll roll this out in this PR and in the subsequent one.

Reply:
In order to do this cleanly on the FIRRTL side, there is now a `class FIRRTLModuleLike : FIRRTLOp<...>` which unifies a lot of common code across `FModuleOp`, `FExtModuleOp`, and `FMemModuleOp`. This was necessary as the `extraModuleClassDefinition` trick only works if `extraClassDefinition` is defined which requires the ODS def/class to either be an op or a type.

This seems to work great! Thanks for the detailed suggestions on how to improve this.